doc/synchronization.md

```diff
@@ -0,0 +1,356 @@
+> Quotations are used for notes.
+
+> This document is work-in-progress.
+> Intentions of this effort and document are: to sum up the behavior
+> of Ruby in concurrent and parallel environment, initiate discussion,
+> identify problems in the document, find flaws in the Ruby
+> implementations if any, suggest what has to be enhanced in Ruby itself
+> and cooperate towards the goal in all implementations (using
+> `concurrent-ruby` as compatibility layer).
+
+# Synchronization
+
+This layer provides tools to write concurrent abstractions independent on any
+particular Ruby implementation. It is build on top of Ruby memory model which is
+also described here. `concurrent-ruby` abstractions are build using this layer.
+
+**Why?** Ruby is great expressive language, but it lacks in support for
+concurrent and parallel computation. It's hoped that this document will provide
+ground steps for Ruby to become as good in this area as in others.
+
+Without memory model and this layer it's very hard to write concurrent
+abstractions for Ruby. To write a proper concurrent abstraction it often means
+to reimplement it more than once for different Ruby runtimes, which is very
+time-consuming and error-prone.
+
+# Ruby memory model
+
+Ruby memory model is a framework allowing to reason about programs in concurrent
+and parallel environment. It allows to identify what is and what is not a [race
+condition](https://en.wikipedia.org/wiki/Race_condition). Memory model is also a
+contract: if a program is written without the race conditions it behaves in
```

Comment:
I'm getting a bit pedantic here, but race conditions are not simply an evil thing. There are several types of race conditions, and here we probably mean data races. Even data races can be part of a well designed application. For example in Go if you have two go-routines and select on them both to get a single value, then that's a race, but that may be what you want.

From the Encylopedia of Parallel Computing (David Padua ed.):

> A race condition occurs in a parallel program execution when two or more threads access a common resource, e.g., a variable in shared memory, and the order of the accesses depends on the timing, i.e., the progress of individual threads.
> The disposition for a race condition is in the parallel program. In different executions of the same program on the same input access events that constitute a race can occur in different order, which may but does not generally result in different program behaviors (non- determinacy).
> Race conditions that can cause unintended non- determinacy are programming errors. Examples of such programming errors are violations of atomicity due to incorrect synchronization.

So a race condition is not an inherently error! It's just a property of the program.